add cgroup v2 support to cgroup_get_current_controller_path()

kamalesh-babulal opened the pull request:
Add support to read/parse cgroup v2 format `/proc/<pid>/cgroup` to
`cgroup_get_current_controller_path()`. The format of `/proc/<pid>/cgroup`
remains the same:
`hierarchy-ID:controller-list:cgroup-path`

the difference is that in cgroup v2 mode:
- `hierarchy-ID` is always 0 (one hierarchy allowed)
- `controller-list` is empty
    
there can also be a mix of both legacy and unified controllers for a
given task if the cgroup mode is hybrid mode.  This patch, add support
for the mixed cgroup setup mode too.

Closes #335

drakenclimber commented on include/libcgroup/groups.h:
```diff
@@ -675,6 +675,12 @@ struct cgroup_controller *cgroup_get_controller_by_index(struct cgroup *cgroup,
+ * Return true if cgroup setup mode is cgroup v2 (unified), else
+ * returns false.
+ */
+bool is_unified_mode(void);
```
nitpick - we probably should remain consistent with previous APIs and use the `cgroup_` prefix.  How does `cgroup_is_unified_mode()` sound?

drakenclimber replied:
Silly question - should we just expose `cgroup_setup_mode()` and give the user even more power.  I'm not sure where you're going with this function, so I could be way off base.

kamalesh-babulal replied:
@drakenclimber Thanks for testing and reviewing the patch.  `cgroup_is_unified_mode()` sounds better, as it's consistent with the other APIs and I will add test cases around it.

> Silly question - should we just expose cgroup_setup_mode() and give the user even more power? I'm not sure where you're going with this function, so I could be way off base.

The idea was to export `is_unified_mode()` as a helper to the users, so they can quickly check if the setup is v1 (legacy/hybrid) or v2.  If it returns true it is v2 or v1 (legacy or hybrid) otherwise.  Exposing `cgroup_setup_mode()` could definitely give more power to the user, but that would also mean exposing `enum cg_setup_mode_t`. 

The next question is, why not have `is_legacy_mode()` and `is_hybrid_mode()` functions?

kamalesh-babulal replied:
`enum cg_setup_mode_t` is already exported, will spin up the next version will `cgroup_setup_mode()` and other helpers.

drakenclimber replied:
I don't have a preference either way.  If you think `cgroup_is_unified_mode()` is easier/better for the user, then I'm cool with that.  If we think giving them the full `cg_setup_mode_t` enum is easier, that's fine, too.

drakenclimber reviewed (commented):
I think these changes look good.  I tested them out on my v2 box, and they behaved exactly as I hoped/expected.

Let's get some tests around this and add it to the 3.1.0 release.  I especially interested in negative tests and how this performs on the hybrid box.  Thanks, @kamalesh-babulal

kamalesh-babulal commented:
V2 Changes:
- Removed the `is_mode_unfiied()` commit, its unrelated to this patch.
- Added test cases for both `legacy/hybrid` and `unified` modes.
- Fix the case of `unified` mode and `controller != NULL`

kamalesh-babulal commented:
**v3 Changes**:
- Rebase the code on top of master branch
- Added the test cases to the `ftests/Makefile.am`
- Added `Reported-by` tag

drakenclimber commented on tests/ftests/081-pybindings-cgrp_get_curr_ctrl_path-v1.py:
```diff
@@ -0,0 +1,133 @@
+    try:
+        cgrp_path = cgrp.get_current_controller_path(pid, "invalid")
+    except RuntimeError as re:
+        if '50011' not in str(re):
```
RuntimeError has access to the return code.  You should be able to do this:
```
if re.ret != 50011:
    raise re
```

kamalesh-babulal replied:
`RunError` has access to `ret`, that exception is raised when we run commands, this via pybindings and trying to check for `re.ret` will error out  with `'RuntimeError' object has no attribute 'ret'`

drakenclimber replied:
D'oh.  You're totally right.  I obviously confused the two :(

drakenclimber commented on tests/ftests/081-pybindings-cgrp_get_curr_ctrl_path-v1.py:
```diff
@@ -0,0 +1,133 @@
+    try:
+        cgrp_path = cgrp.get_current_controller_path(pid, None)
+    except RuntimeError as re:
+        if '50016' not in str(re):
```
Same comment as above

drakenclimber commented on tests/ftests/081-pybindings-cgrp_get_curr_ctrl_path-v1.py:
```diff
@@ -0,0 +1,133 @@
+    result = consts.TEST_PASSED
+    cause = None
+
+    if config.args.container:
```
Out of curiosity, what happens if this is run in a container?

drakenclimber commented on tests/ftests/081-pybindings-cgrp_get_curr_ctrl_path-v1.py:
```diff
@@ -0,0 +1,133 @@
+    result = consts.TEST_PASSED
+    cause = None
+
+    expected_path = os.path.join("/" + CGNAME)
```
nitpick - I don't think `os.path.join()` is doing anything for us here since we're only providing a single string.

drakenclimber commented on src/python/libcgroup.pyx:
```diff
@@ -627,6 +627,40 @@ cdef class Cgroup:
 
         return pid_list
 
+    def get_current_controller_path(self, pid, controller):
```
Also, it doesn't appear that we're using the Cgroup instance at all.  Should this method be made a `@staticmethod`?

kamalesh-babulal replied:
agreed, will make it staticmethod

kamalesh-babulal commented on src/python/libcgroup.pyx:
```diff
@@ -627,6 +627,40 @@ cdef class Cgroup:
 
         return pid_list
 
+    def get_current_controller_path(self, pid, controller):
```
@drakenclimber Thank you so much for the review. Good idea, it will make cgroup v2 better

kamalesh-babulal commented on tests/ftests/081-pybindings-cgrp_get_curr_ctrl_path-v1.py:
```diff
@@ -0,0 +1,133 @@
+    #          It's expected to fail because we not had created cgroup.
+    #
+    cgrp_path = cgrp.get_current_controller_path(pid, "memory")
+    if cgrp_path == expected_path:
```
```
# cgget -m
Legacy Mode (Cgroup v1 only).

# cat /proc/1054/cgroup 
13:pids:/user.slice/user-1000.slice/session-1.scope
12:blkio:/user.slice
11:net_cls,net_prio:/
10:memory:/user.slice/user-1000.slice/session-1.scope
9:perf_event:/
8:cpuset:/
7:misc:/
6:rdma:/
5:hugetlb:/
4:devices:/user.slice
3:cpu,cpuacct:/user.slice
2:freezer:/
1:name=systemd:/user.slice/user-1000.slice/session-1.scope
```

systemd might place it in another scope and not necessarily in '/', I too initially assumed it will be `root` cgroup, so limited the check `expected path`

kamalesh-babulal commented:
**v4 Changes**:
- Addressed the reviews comments from @drakenclimber

drakenclimber commented:
Closes #335

drakenclimber commented:
Pushed to the main branch only.  Thanks @kamalesh-babulal
